lxd/firewall/drivers/drivers_xtables.go

```diff
@@ -965,10 +966,14 @@ func (d Xtables) InstanceSetupProxyNAT(projectName string, instanceName string,
 			return err
 		}
 
-		if targetIndex == i {
+		// Check if 1) we have target ports, 2) the current target port is at the beginning of a port range.
+		if targetPortsLen > 1 && targetRangeIndex < len(forward.TargetPortRanges) && forward.TargetPortRanges[targetRangeIndex][0] == forward.TargetPorts[targetIndex] {
```

Comment:
Now that we are decoupling the MASQUERADE rules from the DNAT rules (by definition that we want to have fewer MASQUERADE rules than we have DNAT rules in certain situations), I think it would make sense to move the MASQUERADE rule setup out of the `for i := range forward.ListenPorts {` loop.

Instead you can move it below, and then just iterate through the TargetPortRanges field I think.